### PR TITLE
Ldap auth

### PR DESCRIPTION
**Pull request by AugustH**
Improved LDAP Authentication v2

- extended config
- allow multiple configs
- documentation

### CODE REVIEW

File: inginious/frontend/plugins/auth/ldap_auth.py [hunk 1]
```diff
@@ -61,7 +61,100 @@ class LDAPAuthenticationPage(AuthenticationPage):
...
+            raise Exception('Unable to contact host')
+        except Exception as e:
+            logger.error('Unable to get user data: {}'.format(str(e)))
+            conn.unbind()
+            raise Exception('Unknown User')
+
+        # 'Log in' (bind) to the ldap connection
+        if conn.rebind(user_data['dn'], password=password):
+            try:
+                email = user_data['attributes'][mail]
```
Comment (anthonygego) [inline inginious/frontend/plugins/auth/ldap_auth.py]: Here we used to take the first item of the returned array. Are the attributes not returned as an array in your case ?

File: inginious/frontend/plugins/auth/ldap_auth.py [hunk 2]
```diff
@@ -61,7 +61,100 @@ class LDAPAuthenticationPage(AuthenticationPage):
...
+                username = login
+                realname = user_data['attributes'][cn]
```
Comment (anthonygego) [inline inginious/frontend/plugins/auth/ldap_auth.py]: Same comment here.

Comment (AugustH) [reply inginious/frontend/plugins/auth/ldap_auth.py]: i see - but i just rechecked against our ldap (windows server): both cn and userPrincipalName (which is used for mail) are strings, not lists. Sadly i can't check the original code as our ldap server needs TLS_BEFORE_BIND. 

Two ways to resolve:
- add [0], which works even in my case, gives bad realnames and mail
- additional `if isinstance` check

Any chance that you/anybody checks the old code: are there really lists or is only the first character taken?

Comment (anthonygego) [reply inginious/frontend/plugins/auth/ldap_auth.py]: I'll check this soon. AFAIK both the LDAP directories we used returned lists. But there may be differences between openLDAP and Active Directory.

**Comment by anthonygego** [comment]
Superseeded by #554

**Pull request closed without merging**